src/SignalR/server/StackExchangeRedis/src/Internal/RedisProtocol.cs

```diff
@@ -104,101 +106,113 @@ public byte[] WriteAck(int messageId)
             // * An 'int': The Id of the command being acknowledged.
             // Any additional items are discarded.
 
-            var writer = MemoryBufferWriter.Get();
+            var memoryBufferWriter = MemoryBufferWriter.Get();
+            var writer = new MessagePackWriter(memoryBufferWriter);
+
             try
             {
-                MessagePackBinary.WriteArrayHeader(writer, 1);
-                MessagePackBinary.WriteInt32(writer, messageId);
+                writer.WriteArrayHeader(1);
+                writer.WriteInt32(messageId);
 
-                return writer.ToArray();
+                return memoryBufferWriter.ToArray();
             }
             finally
             {
-                MemoryBufferWriter.Return(writer);
+                MemoryBufferWriter.Return(memoryBufferWriter);
             }
         }
 
         public RedisInvocation ReadInvocation(ReadOnlyMemory<byte> data)
         {
             // See WriteInvocation for the format
-            ValidateArraySize(ref data, 2, "Invocation");
+            var reader = new MessagePackReader(data);
+            ValidateArraySize(ref reader, 2, "Invocation");
 
             // Read excluded Ids
             IReadOnlyList<string> excludedConnectionIds = null;
-            var idCount = MessagePackUtil.ReadArrayHeader(ref data);
+            var idCount = MessagePackUtil.ReadArrayHeader(ref reader);
             if (idCount > 0)
             {
                 var ids = new string[idCount];
                 for (var i = 0; i < idCount; i++)
                 {
-                    ids[i] = MessagePackUtil.ReadString(ref data);
+                    ids[i] = MessagePackUtil.ReadString(ref reader);
                 }
 
                 excludedConnectionIds = ids;
             }
 
             // Read payload
-            var message = ReadSerializedHubMessage(ref data);
+            // REVIEW : There's a chance that i just created a bug here ...
```

Comment:
Definitely looks like a bug, `data` was being modified before and now the original value of `data` is being passed to `ReadSerializedHubMessage`.

This is an internal class so there are no public APIs to worry about.

Reply:
I only changed `ReadSerializedHubMessage` signature to re-use the existing `MessagePackReader` that is already consuming the same `ReadOnlyMemory<byte> data`

I did not changes `ReadInvocation`, `ReadGroupCommand` or `ReadAck` that also require a `ReadOnlyMemory<byte>`
As they are called from `RedisHubLifetimeManager`, I'm not sure what to think about making the call to `new MessagePackReader()` in the `FooHubLifeTimeManager`

Is that fine for you like this or should i also make this change ?